File: java/circular-buffer/src/main/java/CircularBuffer.java
import java.util.LinkedList;

class CircularBuffer<T> {

    private LinkedList<T> buffer;
    private int capacity;

    CircularBuffer(final int size) {
        this.buffer = new LinkedList<>();
        this.capacity = size;
    }

    T read() throws BufferIOException {
        if (this.buffer.isEmpty()) {
            throw new BufferIOException("Tried to read from empty buffer");
        }

        return this.buffer.poll();
    }

    void write(T data) throws BufferIOException {
        if (this.buffer.size() == this.capacity) {
            throw new BufferIOException("Tried to write to full buffer");
        }

        this.buffer.add(data);
    }

    void overwrite(T data) {
        if (this.buffer.size() == this.capacity) {
            this.buffer.poll();
        }

        this.buffer.add(data);
    }

    void clear() {
        this.buffer.clear();
    }

}

File: python/circular-buffer/circular_buffer.py
class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message =  message


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message = message


class CircularBuffer:
    def __init__(self, capacity):
        self.capacity = capacity
        self.buffer = []

    def read(self):
        if len(self.buffer) == 0:
            raise BufferEmptyException("Circular buffer is empty")

        return self.buffer.pop(0)

    def write(self, data):
        if len(self.buffer) == self.capacity:
            raise BufferFullException("Circular buffer is full")

        self.buffer.append(data)

    def overwrite(self, data):
        if len(self.buffer) == self.capacity:
            self.buffer.pop(0)

        self.buffer.append(data)

    def clear(self):
        self.buffer = []
